src/main/java/teammates/ui/webapi/GetNotificationsAction.java

```diff
@@ -63,11 +64,46 @@ public JsonResult execute() {
                     logic.getActiveNotificationsByTargetUser(targetUser);
         }
 
-        NotificationsData responseData = new NotificationsData(notificationAttributes);
+        String isFetchingAllString = getRequestParamValue(Const.ParamsNames.NOTIFICATION_IS_FETCHING_ALL);
+        if (isFetchingAllString != null
+                && !isFetchingAllString.equals("true") && !isFetchingAllString.equals("false")) {
+            throw new InvalidHttpParameterException(INVALID_IS_FETCHING_ALL);
+        }
+        boolean isFetchingAll = Boolean.parseBoolean(isFetchingAllString);
+
+        if (isFetchingAll) {
+            return new JsonResult(new NotificationsData(notificationAttributes));
+        }
+
+        // Filter unread notifications
+        List<String> readNotifications = logic.getReadNotificationsId(userInfo.getId());
+        notificationAttributes = notificationAttributes
+                .stream()
+                .filter(n -> !readNotifications.contains(n.getNotificationId()))
+                .collect(Collectors.toList());
 
-        if (!userInfo.isAdmin) {
-            responseData.getNotifications().forEach(NotificationData::hideInformationForNonAdmin);
+        if (userInfo.isAdmin) {
+            return new JsonResult(new NotificationsData(notificationAttributes));
+        }
+
+        // Update shown attribute once a non-admin user fetches unread notifications
+        for (NotificationAttributes n : notificationAttributes) {
+            if (n.isShown()) {
+                continue;
+            }
+            try {
+                NotificationAttributes.UpdateOptions newNotification =
+                        NotificationAttributes.updateOptionsBuilder(n.getNotificationId())
+                                .withShown()
+                                .build();
+                logic.updateNotification(newNotification);
+                n.setShown();
```

Comment:
Is this line needed?

Reply:
This is to update the `shown` feature for notification data that returned to the frontend, but since `shown` is not useful for frontend, I think i will remove it first